src/net/ossl.cc

```diff
@@ -385,5 +391,465 @@ void tls::server_credentials::set_client_auth(client_auth ca) {
+                  const auto n = BIO_write(_in_bio, buf->get(), buf->size());
+                  if (n <= 0) {
+                      return make_exception_future<>(ossl_error(n,  "Error while waiting for input"));
+                  }
+                  buf->trim_front(n);
+                  return make_ready_future();
+              }).finally([buf]{});
+        }).handle_exception([](auto ep){
+            return make_exception_future(ep);
+        });
+    }
+
+    future<buf_type> do_get() {
+        // Check if there is encrypted data sitting in ssls internal buffers, otherwise wait
+        // for data and use a
+        auto f = make_ready_future<>();
+        auto avail = BIO_ctrl_pending(_in_bio);
+        if (avail == 0) {
+            if (eof()) {
+                return make_ready_future<buf_type>(buf_type());
+            }
+            f = wait_for_input();
+        }
+        return f.then([this]() {
+            const auto buf_size = 4096;
+            buf_type buf(buf_size);
+            // Read decrypted data from ssls internal buffers
+            auto bytes_read = SSL_read(_ssl.get(), buf.get_write(), buf_size);
+            if (bytes_read <= 0) {
+                const auto ec = SSL_get_error(_ssl.get(), bytes_read);
+                if (ec == SSL_ERROR_ZERO_RETURN && connected()) {
+                    // Client has initiated shutdown by sending EOF
+                    _eof = true;
+                    close();
+                    return make_ready_future<buf_type>(buf_type());
+                }
+                return make_exception_future<buf_type>(ossl_error(ec, "error via SSL_read"));
+            }
+            buf.trim(bytes_read);
+            return make_ready_future<buf_type>(std::move(buf));
+        });
+    }
+
+    future<buf_type> get() override {
+        if (_error) {
+            return make_exception_future<temporary_buffer<char>>(_error);
+        }
+        if (_shutdown || eof()) {
+            return make_ready_future<temporary_buffer<char>>(buf_type());
+        }
+        if (!connected()) {
+            return handshake().then(std::bind(&session::get, this));
+        }
+        return with_semaphore(_in_sem, 1, std::bind(&session::do_get, this)).then([](temporary_buffer<char> buf) {
+            // TODO(rob) - maybe re-handshake?
+            return make_ready_future<temporary_buffer<char>>(std::move(buf));
+        });
+    }
+
+    future<> do_shutdown() {
+        /// TODO(rob) - Is yield the right option here
+        if(_error || !connected()) {
+            return make_ready_future();
+        }
+        auto res = SSL_shutdown(_ssl.get());
+        if (res == 1){
+            // Shutdown has completed successfully
+            return make_ready_future<>();
+        } else if (res == 0) {
+            // Shutdown process is ongoing and has not yet completed, peer has not yet replied
+            return yield().then([this](){
+                return do_shutdown();
+            });
```

Comment:
could this just end up in some infinite loop?

Reply:
Actually it looks like you may need to call SSL_read or something, based on the documentation?

Reply:
look underneath that, condition is handled when SSL_shutdown doesn't return 1 or 0

Reply:
I'm referring to the documentation for [`SSL_shutdown`](https://www.openssl.org/docs/man3.0/man3/SSL_shutdown.html) when it reutrn s0:

> The shutdown is not yet finished: the close_notify was sent but the peer did not send it back yet. Call SSL_read() to do a bidirectional shutdown.
> 
> Unlike most other function, returning 0 does not indicate an error. [SSL_get_error(3)](https://www.openssl.org/docs/man3.0/man3/SSL_get_error.html) should not get called, it may misleadingly indicate an error even though no error occurred.

Reply:
Yeah there's also this, which is in the `nonblocking mode` section:

> After SSL_shutdown() returned 0, it is possible to call SSL_shutdown() again to wait for the peer's close_notify alert. SSL_shutdown() will return 1 in that case. However, it is recommended to wait for it using SSL_read() instead.

Reply:
Maybe calling SSL_read in all cases is cleaner